File: src/lib/camase/imm.ts
import { KalmanIRW, processCov, transition } from "./kalman";

const SCALES = [0.3, 1.0, 3.0];
const N = 3;
const STAY = 0.92;

export function transitionMatrix(stay = STAY): number[][] {
  const off = (1 - stay) / (N - 1);
  return Array.from({ length: N }, (_, i) =>
    Array.from({ length: N }, (_, j) => (i === j ? stay : off)),
  );
}

export class IMMFilter {
  R0: number;
  sigmaA0: number;
  dt: number;
  mu: number[] = [1 / 3, 1 / 3, 1 / 3];
  filters: KalmanIRW[] = [new KalmanIRW(), new KalmanIRW(), new KalmanIRW()];
  Pi = transitionMatrix();
  lastNis = 1;
  lastPred = 0;
  private cbar = [1 / 3, 1 / 3, 1 / 3];

  constructor(R0: number, sigmaA0: number, dt = 1) {
    this.R0 = R0;
    this.sigmaA0 = sigmaA0;
    this.dt = dt;
  }

  get x(): [number, number] {
    return [
      this.mu[0] * this.filters[0].x[0] + this.mu[1] * this.filters[1].x[0] + this.mu[2] * this.filters[2].x[0],
      this.mu[0] * this.filters[0].x[1] + this.mu[1] * this.filters[1].x[1] + this.mu[2] * this.filters[2].x[1],
    ];
  }

  modeR(): number[] {
    return SCALES.map((s) => s * this.R0);
  }
  modeSa(): number[] {
    return SCALES.map((s) => s * this.sigmaA0);
  }

  mix() {
    const cbar = [0, 0, 0];
    for (let j = 0; j < N; j++) {
      for (let i = 0; i < N; i++) cbar[j] += this.Pi[i][j] * this.mu[i];
      cbar[j] = Math.max(cbar[j], 1e-16);
    }
    this.cbar = cbar;
    const xs = this.filters.map((f) => [...f.x] as [number, number]);
    const Ps = this.filters.map((f) => f.P.map((row) => [...row]));
    for (let j = 0; j < N; j++) {
      const xj: [number, number] = [0, 0];
      for (let i = 0; i < N; i++) {
        const w = (this.Pi[i][j] * this.mu[i]) / cbar[j];
        xj[0] += w * xs[i][0];
        xj[1] += w * xs[i][1];
      }
      const Pj = [
        [0, 0],
        [0, 0],
      ];
      for (let i = 0; i < N; i++) {
        const w = (this.Pi[i][j] * this.mu[i]) / cbar[j];
        const d0 = xs[i][0] - xj[0];
        const d1 = xs[i][1] - xj[1];
        Pj[0][0] += w * (Ps[i][0][0] + d0 * d0);
        Pj[0][1] += w * (Ps[i][0][1] + d0 * d1);
        Pj[1][0] += w * (Ps[i][1][0] + d1 * d0);
        Pj[1][1] += w * (Ps[i][1][1] + d1 * d1);
      }
      this.filters[j].x = xj;
      this.filters[j].P = Pj;
    }
  }

  step(y: number): [number, number] {
    const F = transition(this.dt);
    const x = this.x;
    this.lastPred = F[0][0] * x[0] + F[0][1] * x[1];
    this.mix();
    const likes = [0, 0, 0];
    const R = this.modeR();
    const sa = this.modeSa();
    for (let j = 0; j < N; j++) {
      this.filters[j].predict(F, processCov(sa[j], this.dt));
      this.filters[j].updateJoseph(y, R[j]);
      const S = Math.max(this.filters[j].lastS, 1e-18);
      const inn = this.filters[j].lastInnov;
      likes[j] = Math.exp(-0.5 * (inn * inn) / S) / Math.sqrt(2 * Math.PI * S);
    }
    let s = 0;
    const post = this.cbar.map((c, j) => {
      const v = likes[j] * c;
      s += v;
      return v;
    });
    this.mu = s > 0 ? post.map((v) => v / s) : [1 / 3, 1 / 3, 1 / 3];
    let Sm = 0;
    for (let j = 0; j < N; j++) Sm += this.mu[j] * this.filters[j].lastS;
    const innov = y - this.lastPred;
    this.lastNis = (innov * innov) / Math.max(Sm, 1e-18);
    return this.x;
  }
}

File: camase/imm.py
"""3-mode Interacting Multiple Model filter (model M8).

Modes: calm / normal / stress with fixed (R, σ_a²) multiples of the
paper defaults. Mixing is standard Blom / Bar-Shalom. No wavelet on
this path — M8 is a covariance-regime baseline, not a leaky cascade.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .kalman import KalmanIRW, process_cov, transition


MODE_SCALES = (0.3, 1.0, 3.0)
N_MODES = 3
STAY = 0.92


def transition_matrix(stay: float = STAY, n: int = N_MODES) -> np.ndarray:
    off = (1.0 - stay) / (n - 1)
    P = np.full((n, n), off)
    np.fill_diagonal(P, stay)
    return P


@dataclass
class IMMFilter:
    R0: float
    sigma_a0: float
    dt: float = 1.0
    stay: float = STAY
    mu: np.ndarray = field(default_factory=lambda: np.ones(N_MODES) / N_MODES)
    filters: list[KalmanIRW] = field(default_factory=list)
    last_nis: float = 1.0
    last_innov: float = 0.0
    last_pred: float = 0.0

    def __post_init__(self) -> None:
        if not self.filters:
            self.filters = [KalmanIRW() for _ in range(N_MODES)]
        self.Pi = transition_matrix(self.stay)

    @property
    def x(self) -> np.ndarray:
        return sum(m * f.x for m, f in zip(self.mu, self.filters))

    @property
    def mode_R(self) -> tuple[float, ...]:
        return tuple(s * self.R0 for s in MODE_SCALES)

    @property
    def mode_sa(self) -> tuple[float, ...]:
        return tuple(s * self.sigma_a0 for s in MODE_SCALES)

    def mix(self) -> None:
        cbar = self.Pi.T @ self.mu
        cbar = np.maximum(cbar, 1e-16)
        mu_mix = (self.Pi * self.mu[:, None]) / cbar[None, :]
        xs = [f.x.copy() for f in self.filters]
        Ps = [f.P.copy() for f in self.filters]
        mixed_x = []
        mixed_P = []
        for j in range(N_MODES):
            xj = sum(mu_mix[i, j] * xs[i] for i in range(N_MODES))
            Pj = np.zeros((2, 2))
            for i in range(N_MODES):
                d = (xs[i] - xj).reshape(2, 1)
                Pj = Pj + mu_mix[i, j] * (Ps[i] + d @ d.T)
            mixed_x.append(xj)
            mixed_P.append(0.5 * (Pj + Pj.T))
        for j, f in enumerate(self.filters):
            f.x = mixed_x[j]
            f.P = mixed_P[j]
        self._cbar = cbar

    def step(self, y: float) -> np.ndarray:
        F = transition(self.dt)
        self.last_pred = float((F @ self.x)[0])
        self.mix()
        likes = np.zeros(N_MODES)
        for j, f in enumerate(self.filters):
            Q = process_cov(self.mode_sa[j], self.dt)
            f.predict(F, Q)
            f.update_joseph(y, self.mode_R[j])
            S = max(f.last_S, 1e-18)
            likes[j] = np.exp(-0.5 * (f.last_innov ** 2) / S) / np.sqrt(2 * np.pi * S)
        post = likes * self._cbar
        s = post.sum()
        self.mu = post / s if s > 0 else np.ones(N_MODES) / N_MODES
        x = self.x
        innov = y - self.last_pred
        # Combined NIS against mixture innovation variance.
        S_mix = 0.0
        for j, f in enumerate(self.filters):
            S_mix += self.mu[j] * f.last_S
        S_mix = max(S_mix, 1e-18)
        self.last_innov = float(innov)
        self.last_nis = float(innov * innov / S_mix)
        return x
